cmd/argo/commands/submit.go

```diff
@@ -248,3 +271,100 @@ func waitOrWatch(workflowNames []string, cliSubmitOpts cliSubmitOpts) {
 		watchWorkflow(workflowNames[0])
 	}
 }
+
+func parseParameters(opts *util.SubmitOpts) ([]wfv1.Parameter, error) {
+	newParams := make([]wfv1.Parameter, 0)
+	if len(opts.Parameters) > 0 || opts.ParameterFile != "" {
+		passedParams := make(map[string]bool)
+		for _, paramStr := range opts.Parameters {
+			parts := strings.SplitN(paramStr, "=", 2)
+			if len(parts) == 1 {
+				return nil, fmt.Errorf("Expected parameter of the form: NAME=VALUE. Received: %s", paramStr)
+			}
+			param := wfv1.Parameter{
+				Name:  parts[0],
+				Value: &parts[1],
+			}
+			newParams = append(newParams, param)
+			passedParams[param.Name] = true
+		}
+		if opts.ParameterFile != "" {
+			var body []byte
+			var err error
+			if cmdutil.IsURL(opts.ParameterFile) {
+				body, err = util.ReadFromUrl(opts.ParameterFile)
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				body, err = ioutil.ReadFile(opts.ParameterFile)
+				if err != nil {
+					return nil, err
+				}
+			}
+			yamlParams := make(map[string]string)
+			err = yaml.Unmarshal(body, &yamlParams)
+			if err != nil {
+				return nil, err
+			}
+
+			for k, v := range yamlParams {
+				value, err := strconv.Unquote(string(v))
+				if err != nil {
+					value = string(v)
+				}
+				param := wfv1.Parameter{
+					Name:  k,
+					Value: &value,
+				}
+				if _, ok := passedParams[param.Name]; ok {
+					continue
+				}
+				newParams = append(newParams, param)
+				passedParams[param.Name] = true
+			}
+		}
+	}
+	return newParams, nil
+}
+
+func replaceGlobalParameters(fileContents [][]byte, submitOpts *util.SubmitOpts, cliOpts *cliSubmitOpts) ([][]byte, error) {
+	var output [][]byte
+	for _, body := range fileContents {
+		workflowRaw := make(map[interface{}]interface{})
```

Comment:
No need to make a new interface here.
```suggestion
		var workflowRaw wfv1.Workflow
```

Reply:
The issue with using the wfv1.Workflow is that the value for each key then need to have the correct type. For the test where we use the retry limit:
```
retryStrategy:
     limit: "{{workflow.parameters.retry-count}}"
```
 a integer is required but we pass in a string, ```"{{workflow.parameters.retry-count}}" ``` thus this will result in: 

``` 
Failed to parse workflow: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal string into Go struct field RetryStrategy.spec.templates.retryStrategy.limit of type int32
```. 

I might miss something but as I understand this is the reason for why the current implementation where the replacement is done on the server side don't allow this.